src/python/pants/option/parser.py

```diff
@@ -181,9 +182,15 @@ def walk(self, callback: Callable) -> None:
     @frozen_after_init
     @dataclass(unsafe_hash=True)
     class ParseArgsRequest:
-        flag_value_map: Dict
+        # N.B.: We use this callable protool instead of Callable directly to work around the
+        # dataclass-specific issue described here: https://github.com/python/mypy/issues/6910
+        class FlagNameProvider(Protocol):
+            def __call__(self) -> Iterable:
```

Comment:
Even better would be to set the return type as `Sequence["Options.ScopedFlagNameForFuzzyMatching"]`. We used to have this, but I had to remove it when we added file dependencies because we started erroring on a cycle with `options.py`. Now, we tolerate those cycles, so we can add that import back with:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
   import pants.option.options import Options
```

Reply:
Even better would not be to have an un-needed cycle and that's easy. I'll follow up instead of snowballing this PR.

Reply:
Agreed :) Thanks!